src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,191 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        size = size();
+        if (head == null || size == 0) {
+            head = new Node<>(null, null, null);
+            tail = new Node<>(null, null, null);
+            head.next = tail;
+            tail.prev = head;
+        }
+        if (head.value == null) {
+            head.value = value;
+            return;
+        }
+        if (size == 1) {
+            tail.value = value;
+            return;
+        }
+        Node<T> addNewNode = new Node<>(null, value, null);
+        Node<T> currentNode = tail;
+        tail = addNewNode;
+        tail.prev = currentNode;
+        currentNode.next = tail;
     }
 
     @Override
     public void add(T value, int index) {
+        size = size();
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> target = getNode(index);
+        linkBefore(value, target);
+    }
+
+    private void linkBefore(T value, Node<T> newNodeBefore) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (newNodeBefore == head) {
+            newNode.next = newNodeBefore;
+            newNodeBefore.prev = newNode;
+            head = newNode;
+            return;
+        }
+        Node<T> previousNode = newNodeBefore.prev;
+        newNode.next = newNodeBefore;
+        newNodeBefore.prev = newNode;
+        previousNode.next = newNode;
+        newNode.prev = previousNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        exceptionForIndex(index);
+        Node<T> currentNode = getNode(index);
+        return currentNode.value;
```

Comment:
check index inside getNode()
```suggestion
        return getNode(index).value;
```